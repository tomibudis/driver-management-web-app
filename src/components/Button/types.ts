export type Variant = "primary" | "link";

export interface ButtonProps extends React.RefAttributes<HTMLButtonElement> {
  onClick?: () => void;
  rightIcon?: string;
  variant: Variant;
  disabled?: boolean;
}
