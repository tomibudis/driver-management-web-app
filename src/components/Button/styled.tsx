import styled from "styled-components";

import { COLORS } from "~constants/index";
import { ButtonProps } from "./types";

export const StyledButton = styled.button<ButtonProps>`
  display: flex;
  align-items: center;
  background: ${(props) => (props.variant === "primary" ? COLORS.RED : "none")};
  border: none;
  border-radius: 4px;
  padding: 8px 14px;
  width: inherit;
  color: ${(props) => (props.variant === "primary" ? "white" : "black")};
  :hover {
    background: ${(props) => props.variant === "primary" && "#f04a41d4"};
    color: ${(props) => props.variant === "link" && "#000000db"};
  }

  :disabled {
    opacity: 0.5;
  }
`;
