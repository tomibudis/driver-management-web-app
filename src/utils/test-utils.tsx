import React from "react";
import { configureStore, EnhancedStore } from "@reduxjs/toolkit"; // for redux-toolkit
import {
  render as rtlRender,
  RenderOptions,
  RenderResult,
} from "@testing-library/react";
import { ReactElement, ReactNode } from "react";
import { Provider } from "react-redux";
import driversReducer from "../redux/drivers/reducer";

import { RootState } from "../redux/store";
import mockDriversJson from "../views/driver-management/__test__/mock-drivers.json";

type ReduxRenderOptions = {
  preloadedState?: RootState;
  store?: EnhancedStore;
  renderOptions?: Omit<RenderOptions, "wrapper">;
};

function render(
  ui: ReactElement,
  {
    preloadedState = {
      drivers: {
        results: mockDriversJson.results,
        data: mockDriversJson?.results,
      },
    },
    store = configureStore({
      reducer: { drivers: driversReducer },
      preloadedState,
    }),
    ...renderOptions
  }: ReduxRenderOptions = {}
): RenderResult {
  function Wrapper({ children }: { children?: ReactNode }): ReactElement {
    return <Provider store={store}>{children}</Provider>;
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from "@testing-library/react";
export { render };
