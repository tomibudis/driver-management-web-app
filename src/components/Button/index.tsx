import React from "react";
import { StyledButton } from "./styled";
import { ButtonProps } from "./types";

const Button: React.FC<ButtonProps> = ({
  variant = "primary",
  rightIcon,
  onClick,
  children,
  disabled = false,
  ...props
}) => (
  <div>
    <StyledButton
      disabled={disabled}
      variant={variant}
      onClick={onClick}
      {...props}
    >
      {children}
      {rightIcon && <i className={rightIcon} />}
    </StyledButton>
  </div>
);

export default Button;
