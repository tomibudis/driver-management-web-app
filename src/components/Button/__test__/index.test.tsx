import React from "react";
import { COLORS } from "~constants/index";
import { render, screen } from "~utils/test-utils";

import Button from "../index";

describe("<Button />", () => {
  it("render button component", () => {
    render(<Button variant="primary">Button Test</Button>);

    const children = screen.getByText("Button Test");
    expect(children).toBeInTheDocument();
  });

  it("render variant primary", () => {
    render(<Button variant="primary">Button Test</Button>);
    const button = screen.getByText("Button Test");
    expect(button).toHaveStyle(`background: ${COLORS.RED}`);
  });

  it("render variant link", () => {
    render(<Button variant="link">Button Test</Button>);
    const button = screen.getByText("Button Test");
    expect(button).toHaveStyle(`background: none`);
  });

  it("matched with a snapshot", () => {
    const { container } = render(
      <Button variant="primary">Button Test</Button>
    );
    expect(container).toMatchSnapshot();
  });
});
