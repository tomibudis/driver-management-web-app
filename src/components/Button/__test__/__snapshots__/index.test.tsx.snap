// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> matched with a snapshot 1`] = `
<div>
  <button
    class="sc-bczRLJ ibSIfR"
  >
    Button Test
  </button>
</div>
`;
